File: app/src/main/java/com/rex50/mausam/utils/MaterialSnackBar.java
package com.rex50.mausam.utils;

import android.content.Context;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.os.Build;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.IntDef;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.core.content.res.ResourcesCompat;

import com.google.android.material.snackbar.Snackbar;
import com.rex50.mausam.R;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class MaterialSnackBar {
    private Snackbar snackbar;
    private View layout;
    private Context context;

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({LENGTH_INDEFINITE, LENGTH_SHORT, LENGTH_LONG})
    private @interface MaterialSnackBarDuration {
    }

    public static final int LENGTH_INDEFINITE = -2;
    public static final int LENGTH_SHORT = -1;
    public static final int LENGTH_LONG = 0;

    public MaterialSnackBar(Context context, View v) {
        this.context = context;
        this.layout = v;
    }

    public void show(@StringRes int msg, @MaterialSnackBarDuration int duration) {
        show(context.getString(msg), duration, null);
    }

    public void show(String msg, @MaterialSnackBarDuration int duration) {
        show(msg, duration, null);
    }

    public void show(String msg, @MaterialSnackBarDuration int duration, @Nullable Drawable background) {
        if (isSupported()) {
            snackbar = Snackbar.make(layout, msg, duration);
            showSnackBarWithMargin(snackbar, background);
        } else {
            showToast(msg, duration);
        }
    }

    public void showActionSnackBar(@StringRes int msg, @StringRes int action, @MaterialSnackBarDuration int duration, SnackBarListener listener) {
        showActionSnackBar(context.getString(msg), context.getString(action), null, duration, listener);
    }

    public void showActionSnackBar(String msg, String action, @MaterialSnackBarDuration int duration, SnackBarListener listener) {
        showActionSnackBar(msg, action, null, duration, listener);
    }

    public void showActionSnackBar(String msg, String action, @Nullable final Drawable background, @MaterialSnackBarDuration int duration, final SnackBarListener listener) {
        if (isSupported()) {
            snackbar = Snackbar.make(layout, msg, duration)
                    .setAction(action, view -> listener.onActionPressed());
            showSnackBarWithMargin(snackbar, background);
        } else {
            showToast(msg, duration);
        }
    }

    public void dismiss() {
        if(snackbar != null){
            snackbar.dismiss();
            snackbar = null;
        }
    }

    private void showSnackBarWithMargin(Snackbar snackbar, Drawable background) {
        int sideMargin = 28;
        int marginBottom = 28;
        float fontSize = 14;
        Typeface textFont = Typeface.createFromAsset(context.getAssets(), "fonts/Asap-SemiBold.ttf"),
                actionFont = Typeface.createFromAsset(context.getAssets(), "fonts/Asap-Bold.ttf");
        int color = ResourcesCompat.getColor(context.getResources(), R.color.colorAccent, null);
        int colorDark = ResourcesCompat.getColor(context.getResources(), R.color.white, null);
        final View snackBarView = snackbar.getView();
//        TextView textView = (TextView) snackBarView.findViewById(android.support.design.R.id.snackbar_text);
//        textView.setMaxLines(2);

        //setting custom margins
        ViewGroup.MarginLayoutParams params = (ViewGroup.MarginLayoutParams) snackBarView.getLayoutParams();
        params.setMargins(params.leftMargin + sideMargin,
                params.topMargin,
                params.rightMargin + sideMargin,
                params.bottomMargin + marginBottom);

        snackBarView.setLayoutParams(params);
        if (background == null) {
            int radius = 10;
            GradientDrawable drawable = new GradientDrawable();
            drawable.setColor(color);
            drawable.setCornerRadius(radius);
            snackBarView.setBackground(drawable);
        } else
            snackBarView.setBackground(background);


        // change snackbar text color
        /*int snackbarTextId = android.support.design.R.id.snackbar_text;*/  // for support library
        int snackbarTextId = com.google.android.material.R.id.snackbar_text; //for androidx
        int snackbarActionId = com.google.android.material.R.id.snackbar_action;
        TextView textView = snackBarView.findViewById(snackbarTextId);
        textView.setTypeface(textFont);
        textView.setTextColor(colorDark);
        textView.setTextSize(fontSize);
        textView.setMaxLines(5);

        //custom action text color
        textView = snackBarView.findViewById(snackbarActionId);
        textView.setTypeface(actionFont);
        textView.setTextColor(colorDark);
//        snackbar.setActionTextColor(colorDark);

        snackbar.show();
    }

    private boolean isSupported() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP;
    }

    private void showToast(String msg, @MaterialSnackBarDuration int duration) {
        Toast.makeText(context, msg, LENGTH_SHORT == duration ? Toast.LENGTH_SHORT : Toast.LENGTH_LONG).show();
    }

    public interface SnackBarListener {
        void onActionPressed();
    }

}

File: app/src/main/java/com/rex50/mausam/utils/MaterialSnackBar.kt
package com.rex50.mausam.utils

import android.content.Context
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.os.Build
import android.view.View
import android.view.ViewGroup.MarginLayoutParams
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.IntDef
import androidx.annotation.StringRes
import androidx.core.content.res.ResourcesCompat
import com.google.android.material.behavior.SwipeDismissBehavior
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import com.rex50.mausam.R
import com.tingyik90.snackprogressbar.SnackProgressBar
import com.tingyik90.snackprogressbar.SnackProgressBarManager

class MaterialSnackBar(private val context: Context, private val layout: View) {
    private var snackBar: Snackbar? = null
    private val snackProgressBarManager: SnackProgressBarManager? = SnackProgressBarManager(layout, context.lifecycleOwner())
    private var snackProgressBar: SnackProgressBar? = null

    @kotlin.annotation.Retention(AnnotationRetention.SOURCE)
    @IntDef(LENGTH_INDEFINITE, LENGTH_SHORT, LENGTH_LONG)
    private annotation class MaterialSnackBarDuration

    init {
        snackProgressBarManager?.apply {
            useRoundedCornerBackground(true)
            setBackgroundColor(R.color.colorAccent)
            setActionTextColor(R.color.colorAccentDark)
            setProgressBarColor(R.color.white)
            setOverlayLayoutColor(R.color.white_to_dark)
        }
    }

    fun showIndeterminateBar(@StringRes msg: Int){
        snackProgressBar = SnackProgressBar(SnackProgressBar.TYPE_HORIZONTAL, context.getString(msg))
        snackProgressBar?.apply {
            setSwipeToDismiss(false)
            setAllowUserInput(false)
            setIsIndeterminate(true)
        }
        snackProgressBar?.apply {
            snackProgressBarManager?.show(this, SnackProgressBarManager.LENGTH_INDEFINITE)
        }
    }

    fun show(@StringRes msg: Int, @MaterialSnackBarDuration duration: Int) {
        show(context.getString(msg), duration, null)
    }

    @JvmOverloads
    fun show(msg: String, @MaterialSnackBarDuration duration: Int, background: Drawable? = null) {
        if (isSupported) {
            snackBar = Snackbar.make(layout, msg, duration);
            showSnackBarWithMargin(snackBar, background)
        } else {
            showToast(msg, duration)
        }
    }

    fun showActionSnackBar(@StringRes msg: Int, @StringRes action: Int, @MaterialSnackBarDuration duration: Int, listener: SnackBarListener) {
        showActionSnackBar(context.getString(msg), context.getString(action), null, duration, listener)
    }

    fun showActionSnackBar(msg: String, action: String?, @MaterialSnackBarDuration duration: Int, listener: SnackBarListener) {
        showActionSnackBar(msg, action, null, duration, listener)
    }

    fun showActionSnackBar(msg: String, action: String?, background: Drawable?, @MaterialSnackBarDuration duration: Int, listener: SnackBarListener) {
        if (isSupported) {
            snackBar = Snackbar.make(layout, msg, duration)
                    .setAction(action) { view: View? -> listener.onActionPressed() }
            showSnackBarWithMargin(snackBar, background)
        } else {
            showToast(msg, duration)
        }
    }

    fun dismiss() {
        snackBar?.dismiss()
        snackBar = null
        snackProgressBarManager?.dismissAll()
    }

    private fun showSnackBarWithMargin(snackbar: Snackbar?, background: Drawable?) {
        val sideMargin = 0
        val marginBottom = 0
        val fontSize = 14f
        val textFont = Typeface.createFromAsset(context.assets, "fonts/Asap-SemiBold.ttf")
        val actionFont = Typeface.createFromAsset(context.assets, "fonts/Asap-Bold.ttf")
        val color = ResourcesCompat.getColor(context.resources, R.color.colorAccent, null)
        val colorDark = ResourcesCompat.getColor(context.resources, R.color.white, null)
        val snackBarView = snackbar!!.view
        //        TextView textView = (TextView) snackBarView.findViewById(android.support.design.R.id.snackbar_text);
//        textView.setMaxLines(2);

        //setting custom margins
        val params = snackBarView.layoutParams as MarginLayoutParams
        params.setMargins(params.leftMargin + sideMargin,
                params.topMargin,
                params.rightMargin + sideMargin,
                params.bottomMargin + marginBottom)
        snackBarView.layoutParams = params
        if (background == null) {
            val radius = 8
            val drawable = GradientDrawable()
            drawable.setColor(color)
            drawable.cornerRadius = radius.toFloat()
            snackBarView.background = drawable
        } else snackBarView.background = background


        // change snackbar text color
        /*int snackbarTextId = android.support.design.R.id.snackbar_text;*/  // for support library
        val snackbarTextId = com.google.android.material.R.id.snackbar_text //for androidx
        val snackbarActionId = com.google.android.material.R.id.snackbar_action
        var textView = snackBarView.findViewById<TextView>(snackbarTextId)
        textView.typeface = textFont
        textView.setTextColor(colorDark)
        textView.textSize = fontSize
        textView.maxLines = 5

        //custom action text color
        textView = snackBarView.findViewById(snackbarActionId)
        textView.typeface = actionFont
        textView.setTextColor(colorDark)
        //        snackbar.setActionTextColor(colorDark);
        val baseTransientBottomBar: BaseTransientBottomBar<Snackbar>? = snackbar
        baseTransientBottomBar!!.animationMode = BaseTransientBottomBar.ANIMATION_MODE_FADE
        baseTransientBottomBar.show()

        //snackbar.show();
    }

    private val isSupported: Boolean = Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP

    private fun showToast(msg: String, @MaterialSnackBarDuration duration: Int) {
        Toast.makeText(context, msg, if (LENGTH_SHORT == duration) Toast.LENGTH_SHORT else Toast.LENGTH_LONG).show()
    }

    interface SnackBarListener {
        fun onActionPressed()
    }

    companion object {
        const val LENGTH_INDEFINITE = -2
        const val LENGTH_SHORT = -1
        const val LENGTH_LONG = 0
    }

}